composer/trainer/trainer.py

```diff
@@ -750,6 +750,24 @@ def __init__(
         if num_optimizers != 1:
             raise NotImplementedError(f"Only one optimizer is supported; found {num_optimizers} optimizers")
 
+        # Move the model and optimizers to the device
+        if not deepspeed_enabled:
+            # DeepSpeed handles the movement and sharding automatically
+            host_model_params = model.parameters()
+            model = self._device.module_to_device(model)
+            device_model_params = model.parameters()
+
+            # Use surgery to update the parameters of the optimizers, now that the model is on the device
+            # see https://pytorch.org/docs/stable/optim.html#constructing-it
+            module_surgery.replace_params_in_optimizer(
+                old_params=host_model_params,
+                new_params=device_model_params,
+                optimizers=optimizers,
+            )
+
+            # Move any remaining optimizer parameters onto the device
+            optimizers = map_collection(optimizers, self._device.optimizer_to_device)
```

Comment:
It is possible that some internal optimizer buffers are created on CPU. Line 802 moves those parameters over. I elaborated on the comment above to make this clearer why we have this call. (Also, since `replace_params_in_optimizer` is removed, hopefully that makes it less confusing)